File: daemon/internal/livevalidation/attempts.go
package livevalidation

import (
	"context"
	"fmt"
	"time"

	"github.com/dopejs/dope-agent/daemon/internal/identity"
	"github.com/dopejs/dope-agent/daemon/internal/tenantctx"
)

func (m *Manager) newAttempt(input StartInput, tenantContext identity.TenantContext, now time.Time) Attempt {
	validationID := input.ValidationID
	if validationID == "" {
		validationID = newID("live_validation")
	}
	scope := input.RequestedScope
	scope.ValidationID = firstNonEmpty(scope.ValidationID, validationID)
	return Attempt{
		ValidationID:       validationID,
		TenantID:           tenantContext.TenantID,
		CandidateID:        input.CandidateID,
		SourceAttemptID:    input.SourceAttemptID,
		RequestedBy:        tenantContext.PrincipalID,
		EnvironmentScope:   m.EnvironmentScope(),
		RequestedScope:     scope,
		Status:             AttemptStatusQueued,
		PermissionDecision: GateDecision{Allowed: true, CheckedAt: now},
		QuotaDecision:      GateDecision{Allowed: true, CheckedAt: now},
		KillSwitchDecision: GateDecision{Allowed: true, CheckedAt: now},
		ApprovalSummary:    ApprovalSummary{},
		LedgerSummary:      LedgerSummary{},
		CreatedAt:          now,
		UpdatedAt:          now,
	}
}

func (m *Manager) block(ctx context.Context, attempt Attempt, gate, reasonCode, message, reference string) (StartResult, error) {
	attempt.Status = AttemptStatusBlocked
	attempt.UpdatedAt = m.clock()
	if err := m.persistAttempt(ctx, attempt); err != nil {
		return StartResult{}, err
	}
	denial := Denial{Gate: gate, ReasonCode: reasonCode, Message: message, Reference: reference}
	return StartResult{Attempt: attempt, Denials: []Denial{denial}}, ErrLiveValidationBlocked
}

func (m *Manager) persistAttempt(ctx context.Context, attempt Attempt) error {
	if m.store == nil {
		return nil
	}
	return m.store.UpsertLiveValidationAttempt(ctx, attempt)
}

func (m *Manager) ListAttempts(ctx context.Context, filter AttemptFilter) ([]Attempt, error) {
	if m.store == nil {
		return nil, nil
	}
	filter.EnvironmentScope = firstNonEmpty(filter.EnvironmentScope, m.EnvironmentScope())
	return m.store.ListLiveValidationAttempts(ctx, filter)
}

func (m *Manager) GetAttempt(ctx context.Context, validationID string) (Attempt, bool, error) {
	if m.store == nil {
		return Attempt{}, false, nil
	}
	tenantID := ""
	if tenantContext, ok := tenantctx.FromContext(ctx); ok {
		tenantID = tenantContext.TenantID
	}
	return m.store.GetLiveValidationAttempt(ctx, tenantID, validationID)
}

func (m *Manager) Abort(ctx context.Context, validationID string) (Attempt, error) {
	attempt, ok, err := m.GetAttempt(ctx, validationID)
	if err != nil {
		return Attempt{}, err
	}
	if !ok {
		return Attempt{}, fmt.Errorf("live validation %s not found", validationID)
	}
	if attempt.Status == AttemptStatusCompleted || attempt.Status == AttemptStatusAborted || attempt.Status == AttemptStatusFailed {
		return attempt, nil
	}
	now := m.clock()
	attempt.Status = AttemptStatusAborted
	attempt.CompletedAt = &now
	attempt.UpdatedAt = now
	if err := m.persistAttempt(ctx, attempt); err != nil {
		return Attempt{}, err
	}
	return attempt, nil
}

File: crates/domains/livevalidation/src/attempts.rs
//! Attempt lifecycle (port of `attempts.go`): construction, blocking, and
//! the list/get/abort operations.

use chrono::DateTime;
use chrono::Utc;

use kura_identity::TenantContext;
use kura_identity::tenantctx;

use crate::error::LiveValidationError;
use crate::error::StartFailure;
use crate::manager::Denial;
use crate::manager::Manager;
use crate::manager::StartInput;
use crate::manager::StartResult;
use crate::manager::first_non_empty;
use crate::store::AttemptFilter;
use crate::types::Attempt;
use crate::types::AttemptStatus;
use crate::types::GateDecision;

impl Manager {
    /// Port of `Manager.newAttempt`.
    pub(crate) fn new_attempt(
        &self,
        input: &StartInput,
        tenant_context: &TenantContext,
        now: DateTime<Utc>,
    ) -> Attempt {
        let mut validation_id = input.validation_id.clone();
        if validation_id.is_empty() {
            validation_id = crate::manager::new_id("live_validation");
        }
        let mut scope = input.requested_scope.clone();
        scope.validation_id =
            first_non_empty([scope.validation_id.as_str(), validation_id.as_str()]);
        Attempt {
            validation_id,
            tenant_id: tenant_context.tenant_id.clone(),
            candidate_id: input.candidate_id.clone(),
            source_attempt_id: input.source_attempt_id.clone(),
            requested_by: tenant_context.principal_id.clone(),
            environment_scope: self.environment_scope().to_string(),
            requested_scope: scope,
            status: AttemptStatus::from(AttemptStatus::QUEUED),
            permission_decision: GateDecision {
                allowed: true,
                checked_at: now,
                ..GateDecision::default()
            },
            quota_decision: GateDecision {
                allowed: true,
                checked_at: now,
                ..GateDecision::default()
            },
            kill_switch_decision: GateDecision {
                allowed: true,
                checked_at: now,
                ..GateDecision::default()
            },
            created_at: now,
            updated_at: now,
            ..Attempt::default()
        }
    }

    /// Port of `Manager.block`: persists the blocked attempt and returns the
    /// `StartFailure::Blocked` verdict carrying it.
    pub(crate) async fn block(
        &self,
        mut attempt: Attempt,
        gate: &str,
        reason_code: &str,
        message: &str,
        reference: &str,
    ) -> Result<StartFailure, LiveValidationError> {
        attempt.status = AttemptStatus::from(AttemptStatus::BLOCKED);
        attempt.updated_at = self.now();
        self.persist_attempt(&attempt).await?;
        let denial = Denial {
            gate: gate.to_string(),
            reason_code: reason_code.to_string(),
            message: message.to_string(),
            reference: reference.to_string(),
        };
        Ok(StartFailure::Blocked(StartResult {
            attempt,
            denials: vec![denial],
        }))
    }

    /// Port of `Manager.persistAttempt`.
    pub(crate) async fn persist_attempt(
        &self,
        attempt: &Attempt,
    ) -> Result<(), LiveValidationError> {
        if let Some(store) = self.store() {
            store.upsert_attempt(attempt.clone()).await?;
        }
        Ok(())
    }

    /// Port of `Manager.ListAttempts`.
    pub async fn list_attempts(
        &self,
        mut filter: AttemptFilter,
    ) -> Result<Vec<Attempt>, LiveValidationError> {
        let Some(store) = self.store() else {
            return Ok(Vec::new());
        };
        filter.environment_scope =
            first_non_empty([filter.environment_scope.as_str(), self.environment_scope()]);
        store.list_attempts(filter).await
    }

    /// Port of `Manager.GetAttempt`.
    pub async fn get_attempt(
        &self,
        validation_id: &str,
    ) -> Result<Option<Attempt>, LiveValidationError> {
        let Some(store) = self.store() else {
            return Ok(None);
        };
        let tenant_id = tenantctx::from_context()
            .map(|ctx| ctx.tenant_id)
            .unwrap_or_default();
        store.get_attempt(&tenant_id, validation_id).await
    }

    /// Port of `Manager.Abort`.
    pub async fn abort(&self, validation_id: &str) -> Result<Attempt, LiveValidationError> {
        let Some(mut attempt) = self.get_attempt(validation_id).await? else {
            return Err(LiveValidationError::NotFound(validation_id.to_string()));
        };
        let terminal = [
            AttemptStatus::from(AttemptStatus::COMPLETED),
            AttemptStatus::from(AttemptStatus::ABORTED),
            AttemptStatus::from(AttemptStatus::FAILED),
        ];
        if terminal.contains(&attempt.status) {
            return Ok(attempt);
        }
        let now = self.now();
        attempt.status = AttemptStatus::from(AttemptStatus::ABORTED);
        attempt.completed_at = Some(now);
        attempt.updated_at = now;
        self.persist_attempt(&attempt).await?;
        Ok(attempt)
    }
}
